File: CodingExercise/ObjectiveC/MasterCollectionCells.m
//
//  MasterCollectionCells.m
//  SimpsonsViewerOC
//
//  Created by Samuel Maffei on 7/1/18.
//  Copyright © 2018 sammaffei. All rights reserved.
//

#import "MasterCollectionCells.h"
#import <SDWebImage/UIImageView+WebCache.h>

// ------------------------------------------------------------------------------------------------

@implementation SelectableColCell   // Implement to satify compiler

@end

// ------------------------------------------------------------------------------------------------

@implementation TextOnlyColCell

+(NSString *)cellIndentifier
    {
    return @"TextOnlyCell";
    }

-(void)setData:(ItemData *)inData
    {
    _textLabel.text = (inData == nil) ? nil : [inData title];
    }

- (void)setSelected:(BOOL)selected {
    
    self.contentView.backgroundColor = selected ? self.selectedColor : self.unselectedColor;
    
    super.selected = selected;
}

@end

// ------------------------------------------------------------------------------------------------

@implementation IconColCell

+(NSString *)cellIndentifier
    {
    return @"IconCell";
    }

-(void)setData:(ItemData *)inData
    {
    if ((inData == nil) || (inData.imageURL == nil))
        {
        _imageView.image = [UIImage imageNamed:@"MissingImage"];
        }
    else
        {
        [_imageView sd_setImageWithURL:inData.imageURL];
        }
    }

- (void)setSelected:(BOOL)selected
    {
    _imageView.backgroundColor = selected ? self.selectedColor : self.unselectedColor;
    
    super.selected = selected;
    }


@end

File: CodingExercise/MasterCollectionCells.swift
//
//  MasterCollectionCells.swift
//  CodingExercise
//
//  Created by Samuel Maffei on 6/19/18.
//  Copyright © 2018 xfinity. All rights reserved.
//

import UIKit

// Since all the cells set stuff with the ItemData data structure, we can use a protocol
// to genericize setting up cells

protocol SetDataProtocol
    {
    static var cellIdentifer: String { get }
    
    func setData(inData : MasterCollectionViewController.ItemData)
    }

class SelectableColCell : UICollectionViewCell
    {
    @IBInspectable fileprivate var unselectedColor : UIColor!       // Make these inspectable so that you can
    @IBInspectable fileprivate var selectedColor : UIColor!         // set them in interfacebuilder
    }

class TextOnlyColCell : SelectableColCell, SetDataProtocol
    {
    @IBOutlet weak private var textLabel : UILabel!
    
    static var cellIdentifer = "TextOnlyCell"
    
    override var isSelected: Bool{
        didSet
            {
            self.contentView.backgroundColor = self.isSelected ? selectedColor : unselectedColor
            }
    }
    
    func setData(inData : MasterCollectionViewController.ItemData)
        {
        textLabel.text = inData.title
        }
    }

class IconColCell : SelectableColCell, SetDataProtocol
    {
    @IBOutlet weak private var imageView : UIImageView!
    
    static var cellIdentifer = "IconCell"
    
    override var isSelected: Bool{
        didSet
            {
            imageView.backgroundColor = self.isSelected ? selectedColor : unselectedColor
            }
        }

    
    func setData(inData : MasterCollectionViewController.ItemData)
        {
        if let imgUrl = inData.imageURL
            {
            imageView.sd_setImage(with: imgUrl, completed: nil)
            }
        else
            {
            imageView.image = UIImage(named: "MissingImage")
            }
        }
    }
